File: android-app/app/src/main/java/com/mianu/paymentapp/network/ApiService.kt
package com.mianu.paymentapp.network

import com.mianu.paymentapp.data.models.*
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path
import retrofit2.http.Query

/**
 * Retrofit surface for the FastAPI backend.
 *
 * Paths here are written to match the server's routes exactly, with no trailing slashes. FastAPI
 * would answer `users/` with a 307 to `users`, and while OkHttp does preserve the method across a
 * 307, relying on a redirect for every write is a needless round trip.
 */
interface ApiService {

    /** Leading slash: health sits at the host root, outside the `/api/v1` base path. */
    @GET("/health")
    suspend fun checkHealth(): Map<String, String>

    // region Auth

    @POST("auth/login")
    suspend fun login(@Body request: LoginRequest): TokenPair

    @POST("auth/register")
    suspend fun register(@Body request: RegisterRequest): TokenPair

    // endregion

    // region Dashboard

    @GET("dashboard/overview")
    suspend fun getDashboardOverview(): DashboardOverview

    // endregion

    // region Users

    @GET("users")
    suspend fun getUsers(): List<UserResponse>

    @POST("users")
    suspend fun createUser(@Body user: UserCreate): UserResponse

    @GET("users/{user_id}")
    suspend fun getUser(@Path("user_id") userId: String): UserResponse

    @PUT("users/{user_id}")
    suspend fun updateUser(
        @Path("user_id") userId: String,
        @Body update: UserUpdate,
    ): UserResponse

    @DELETE("users/{user_id}")
    suspend fun deleteUser(@Path("user_id") userId: String): Map<String, Any>

    @POST("users/{user_id}/meals/adjust")
    suspend fun adjustUserMeals(
        @Path("user_id") userId: String,
        @Body body: MealAdjustRequest,
    ): UserResponse

    /** Server route is `users/cards/link` — an earlier client used `users/link-card`, which 404s. */
    @POST("users/cards/link")
    suspend fun linkCard(@Body request: CardLinkRequest): CardResponse

    @GET("users/cards/{uid}")
    suspend fun getCard(@Path("uid") uid: String): CardResponse

    // endregion

    // region Meals

    @POST("meals/swipe")
    suspend fun swipeMeal(@Body request: MealSwipeRequest): MealSwipeResponse

    // endregion

    // region Transactions

    @POST("transactions")
    suspend fun createTransaction(@Body request: TransactionRequest): TransactionResponse

    // endregion

    // region Teams

    @GET("teams")
    suspend fun getTeams(): List<TeamResponse>

    @POST("teams")
    suspend fun createTeam(@Body team: TeamCreate): TeamResponse

    @PUT("teams/{team_id}")
    suspend fun updateTeam(
        @Path("team_id") teamId: String,
        @Body update: TeamUpdate,
    ): TeamResponse

    @DELETE("teams/{team_id}")
    suspend fun deleteTeam(@Path("team_id") teamId: String): Map<String, Any>

    @POST("teams/{team_id}/members/{user_id}")
    suspend fun assignMember(
        @Path("team_id") teamId: String,
        @Path("user_id") userId: String,
        @Query("role") role: TeamRole,
    ): Map<String, String>

    @DELETE("teams/{team_id}/members/{user_id}")
    suspend fun removeMemberFromTeam(
        @Path("team_id") teamId: String,
        @Path("user_id") userId: String,
    ): Map<String, Any>

    /** `sender_id` is a required query param server-side; omitting it fails validation with a 422. */
    @POST("teams/{team_id}/notifications")
    suspend fun createTeamNotification(
        @Path("team_id") teamId: String,
        @Query("sender_id") senderId: String,
        @Body notification: NotificationCreate,
    ): NotificationResponse

    @GET("teams/{team_id}/notifications")
    suspend fun getTeamNotifications(
        @Path("team_id") teamId: String,
    ): List<NotificationResponse>

    @GET("teams/{team_id}/members")
    suspend fun getTeamMembers(
        @Path("team_id") teamId: String,
    ): List<UserResponse>

    /** Targeted broadcast. Supersedes [createTeamNotification], which only ever hit one team. */
    @POST("notifications")
    suspend fun broadcast(
        @Query("sender_id") senderId: String,
        @Body body: BroadcastCreate,
    ): NotificationResponse

    @GET("notifications")
    suspend fun getNotifications(
        @Query("sender_id") senderId: String? = null,
        @Query("team_id") teamId: String? = null,
        @Query("limit") limit: Int = 50,
    ): List<NotificationResponse>

    @GET("notifications/inbox/{user_id}")
    suspend fun getInbox(
        @Path("user_id") userId: String,
    ): List<NotificationResponse>

    // endregion

    // region Access control / halls

    @GET("access/halls")
    suspend fun getHalls(): List<HallResponse>

    /** Identity-free room-finding view, for non-staff roles. */
    @GET("access/halls/availability")
    suspend fun getHallAvailability(): List<HallAvailability>

    @POST("access/halls")
    suspend fun createHall(@Body hall: HallCreate): HallResponse

    @PUT("access/halls/{hall_id}")
    suspend fun updateHall(
        @Path("hall_id") hallId: String,
        @Body update: HallUpdate,
    ): HallResponse

    /** Returns the user ids currently inside the hall. */
    @GET("access/halls/{hall_id}/presence")
    suspend fun getHallPresence(@Path("hall_id") hallId: String): List<String>

    /** Body, not query params — the server binds an `AccessScanRequest`. */
    @POST("access/scan")
    suspend fun scanAccess(@Body request: AccessScanRequest): AccessLogResponse

    // endregion

    // region Analytics & reports

    @GET("analytics/daily")
    suspend fun getDailyAnalytics(): AnalyticsSummary

    @GET("analytics/trends")
    suspend fun getTrendAnalytics(): AnalyticsSummary

    @GET("reports/meals")
    suspend fun getMealsReport(): ReportSummary

    @GET("reports/transactions")
    suspend fun getTransactionsReport(): ReportSummary

    @GET("reports/users")
    suspend fun getUsersReport(): ReportSummary

    @GET("reports/export")
    suspend fun exportReport(): Map<String, String>

    // endregion

    // region Settings

    @PUT("settings/meal-windows")
    suspend fun updateMealWindows(
        @Body request: MealWindowRequest,
    ): Map<String, MealWindowRequest>

    // endregion

    // region Operations

    @GET("operations/locations")
    suspend fun getLocations(): List<DelegateLocation>

    @GET("operations/scans")
    suspend fun getLiveScans(
        @Query("type") type: String? = null,
    ): List<LiveScanEvent>

    // endregion

    // region Conference Attendance

    @GET("attendance")
    suspend fun getAttendance(): AttendanceResponse

    @POST("attendance/swipe")
    suspend fun recordAttendanceSwipe(
        @Body request: AttendanceSwipeRequest,
    ): AttendanceSwipeResponse

    @POST("attendance/single")
    suspend fun updateSingleAttendance(
        @Body request: SingleAttendanceRequest,
    ): SingleAttendanceResponse

    @POST("attendance/bulk")
    suspend fun updateBulkAttendance(
        @Body request: BulkAttendanceRequest,
    ): BulkAttendanceResponse

    // endregion
}

File: ios-app/MIANU/Services/APIService.swift
import Foundation

public final class APIService {
    public static let shared = APIService()
    private let client = APIClient.shared
    private let encoder = JSONEncoder()
    
    private init() {}
    
    // MARK: - Health
    public func checkHealth() async throws -> [String: String] {
        try await client.request(endpoint: "/health", method: "GET")
    }
    
    // MARK: - Auth
    public func login(request: LoginRequest) async throws -> TokenPair {
        let body = try encoder.encode(request)
        return try await client.request(endpoint: "auth/login", method: "POST", body: body)
    }
    
    public func register(request: RegisterRequest) async throws -> TokenPair {
        let body = try encoder.encode(request)
        return try await client.request(endpoint: "auth/register", method: "POST", body: body)
    }
    
    // MARK: - Dashboard
    public func getDashboardOverview(token: String) async throws -> DashboardOverview {
        try await client.request(endpoint: "dashboard/overview", method: "GET", token: token)
    }
    
    // MARK: - Users
    public func getUsers(token: String) async throws -> [UserResponse] {
        try await client.request(endpoint: "users", method: "GET", token: token)
    }
    
    public func createUser(user: UserCreate, token: String) async throws -> UserResponse {
        let body = try encoder.encode(user)
        return try await client.request(endpoint: "users", method: "POST", body: body, token: token)
    }
    
    public func getUser(userId: String, token: String) async throws -> UserResponse {
        try await client.request(endpoint: "users/\(userId)", method: "GET", token: token)
    }
    
    public func updateUser(userId: String, update: UserUpdate, token: String) async throws -> UserResponse {
        let body = try encoder.encode(update)
        return try await client.request(endpoint: "users/\(userId)", method: "PUT", body: body, token: token)
    }
    
    public func deleteUser(userId: String, token: String) async throws -> EmptyResponse {
        try await client.request(endpoint: "users/\(userId)", method: "DELETE", token: token)
    }
    
    public func adjustUserMeals(userId: String, body: MealAdjustRequest, token: String) async throws -> UserResponse {
        let reqBody = try encoder.encode(body)
        return try await client.request(endpoint: "users/\(userId)/meals/adjust", method: "POST", body: reqBody, token: token)
    }
    
    public func linkCard(request: CardLinkRequest, token: String) async throws -> CardResponse {
        let body = try encoder.encode(request)
        return try await client.request(endpoint: "users/cards/link", method: "POST", body: body, token: token)
    }
    
    public func getCard(uid: String, token: String) async throws -> CardResponse {
        try await client.request(endpoint: "users/cards/\(uid)", method: "GET", token: token)
    }
    
    // MARK: - Meals
    public func swipeMeal(request: MealSwipeRequest, token: String) async throws -> MealSwipeResponse {
        let body = try encoder.encode(request)
        return try await client.request(endpoint: "meals/swipe", method: "POST", body: body, token: token)
    }
    
    // MARK: - Attendance
    public func getAttendance(token: String) async throws -> AttendanceResponse {
        try await client.request(endpoint: "attendance", method: "GET", token: token)
    }
    
    public func recordAttendanceSwipe(request: AttendanceSwipeRequest, token: String) async throws -> AttendanceSwipeResponse {
        let body = try encoder.encode(request)
        return try await client.request(endpoint: "attendance/swipe", method: "POST", body: body, token: token)
    }
    
    public func updateSingleAttendance(request: SingleAttendanceRequest, token: String) async throws -> SingleAttendanceResponse {
        let body = try encoder.encode(request)
        return try await client.request(endpoint: "attendance/single", method: "POST", body: body, token: token)
    }
    
    public func updateBulkAttendance(request: BulkAttendanceRequest, token: String) async throws -> BulkAttendanceResponse {
        let body = try encoder.encode(request)
        return try await client.request(endpoint: "attendance/bulk", method: "POST", body: body, token: token)
    }
    
    // MARK: - Locations & Operations
    public func getLocations(token: String) async throws -> [DelegateLocation] {
        try await client.request(endpoint: "operations/locations", method: "GET", token: token)
    }
    
    public func getLiveScans(type: String? = nil, token: String) async throws -> [LiveScanEvent] {
        var queryItems: [URLQueryItem]? = nil
        if let type = type {
            queryItems = [URLQueryItem(name: "type", value: type)]
        }
        return try await client.request(endpoint: "operations/scans", method: "GET", token: token, queryItems: queryItems)
    }
    
    // MARK: - Access Control & Halls
    public func getHalls(token: String) async throws -> [HallResponse] {
        try await client.request(endpoint: "access/halls", method: "GET", token: token)
    }
    
    public func getHallAvailability(token: String) async throws -> [HallAvailability] {
        try await client.request(endpoint: "access/halls/availability", method: "GET", token: token)
    }
    
    public func createHall(hall: HallCreate, token: String) async throws -> HallResponse {
        let body = try encoder.encode(hall)
        return try await client.request(endpoint: "access/halls", method: "POST", body: body, token: token)
    }
    
    public func updateHall(hallId: String, update: HallUpdate, token: String) async throws -> HallResponse {
        let body = try encoder.encode(update)
        return try await client.request(endpoint: "access/halls/\(hallId)", method: "PUT", body: body, token: token)
    }
    
    public func getHallPresence(hallId: String, token: String) async throws -> [String] {
        try await client.request(endpoint: "access/halls/\(hallId)/presence", method: "GET", token: token)
    }
    
    public func scanAccess(request: AccessScanRequest, token: String) async throws -> AccessLogResponse {
        let body = try encoder.encode(request)
        return try await client.request(endpoint: "access/scan", method: "POST", body: body, token: token)
    }
    
    // MARK: - Teams
    public func getTeams(token: String) async throws -> [TeamResponse] {
        try await client.request(endpoint: "teams", method: "GET", token: token)
    }
    
    public func createTeam(team: TeamCreate, token: String) async throws -> TeamResponse {
        let body = try encoder.encode(team)
        return try await client.request(endpoint: "teams", method: "POST", body: body, token: token)
    }
    
    public func updateTeam(teamId: String, update: TeamUpdate, token: String) async throws -> TeamResponse {
        let body = try encoder.encode(update)
        return try await client.request(endpoint: "teams/\(teamId)", method: "PUT", body: body, token: token)
    }
    
    public func deleteTeam(teamId: String, token: String) async throws -> EmptyResponse {
        try await client.request(endpoint: "teams/\(teamId)", method: "DELETE", token: token)
    }
    
    public func assignMember(teamId: String, userId: String, role: TeamRole, token: String) async throws -> [String: String] {
        let query = [URLQueryItem(name: "role", value: role.rawValue)]
        return try await client.request(endpoint: "teams/\(teamId)/members/\(userId)", method: "POST", token: token, queryItems: query)
    }
    
    public func removeMember(teamId: String, userId: String, token: String) async throws -> EmptyResponse {
        try await client.request(endpoint: "teams/\(teamId)/members/\(userId)", method: "DELETE", token: token)
    }
    
    public func getTeamMembers(teamId: String, token: String) async throws -> [UserResponse] {
        try await client.request(endpoint: "teams/\(teamId)/members", method: "GET", token: token)
    }
    
    // MARK: - Notifications & Broadcasts
    public func broadcast(senderId: String, body: BroadcastCreate, token: String) async throws -> NotificationResponse {
        let reqBody = try encoder.encode(body)
        let query = [URLQueryItem(name: "sender_id", value: senderId)]
        return try await client.request(endpoint: "notifications", method: "POST", body: reqBody, token: token, queryItems: query)
    }
    
    public func getNotifications(limit: Int = 50, token: String) async throws -> [NotificationResponse] {
        let query = [URLQueryItem(name: "limit", value: "\(limit)")]
        return try await client.request(endpoint: "notifications", method: "GET", token: token, queryItems: query)
    }
    
    public func getInbox(userId: String, token: String) async throws -> [NotificationResponse] {
        try await client.request(endpoint: "notifications/inbox/\(userId)", method: "GET", token: token)
    }
    
    // MARK: - Analytics & Reports
    public func getDailyAnalytics(token: String) async throws -> AnalyticsSummary {
        try await client.request(endpoint: "analytics/daily", method: "GET", token: token)
    }
    
    public func getTrendAnalytics(token: String) async throws -> AnalyticsSummary {
        try await client.request(endpoint: "analytics/trends", method: "GET", token: token)
    }
    
    public func getMealsReport(token: String) async throws -> ReportSummary {
        try await client.request(endpoint: "reports/meals", method: "GET", token: token)
    }
    
    public func getTransactionsReport(token: String) async throws -> ReportSummary {
        try await client.request(endpoint: "reports/transactions", method: "GET", token: token)
    }
    
    public func getUsersReport(token: String) async throws -> ReportSummary {
        try await client.request(endpoint: "reports/users", method: "GET", token: token)
    }
    
    public func exportReport(token: String) async throws -> [String: String] {
        try await client.request(endpoint: "reports/export", method: "GET", token: token)
    }
    
    // MARK: - Settings
    public func updateMealWindows(request: MealWindowRequest, token: String) async throws -> [String: MealWindowRequest] {
        let body = try encoder.encode(request)
        return try await client.request(endpoint: "settings/meal-windows", method: "PUT", body: body, token: token)
    }
}
